{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Режет файл с разметкой (GenBank'a) на файлы по хромосомам"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess \n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "делает из GTF с разметкой датафрейм, упорядочивает его по хромосомам и по началу-концу фичей, ищет бинарным поском границы хромосом и режет по ним. Полученное пишется в файлы по хромосомам\n",
    "\n",
    "\n",
    "(можно проще, но когда писалось я об этом не знал)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subprocess.call('rsync rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_feature_table.txt.gz ~/Documents/CourseWork/genomes/Annotation/feature_table_38.txt.gz', shell=True)\n",
    "#subprocess.call(['gunzip ~/Documents/CourseWork/genomes/Annotation/*.gz'], shell=True)\n",
    "#subprocess.call(['sort -t$\"\\t\" -k2 feature_table_38.txt > sort_feature_table_38.tsv'], shell=True)\n",
    "#subprocess.call(['rm feature_table_38.txt'], shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "feature_table = read_gtf(\"/Users/pitikov_egor/Documents/CourseWork/genomes/Annotation/gencode.v35.annotation.gtf\")\n",
    "feature_table.sort_values(by=['seqname', 'start', 'end'], kind='mergesort').reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#########subprocess.call(['rm /Users/pitikov_egor/Documents/CourseWork/genomes/Annotation/gencode.v35.annotation.gtf'], shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_table = pd.read_csv('/Users/pitikov_egor/Documents/CourseWork/genomes/Annotation/genes.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60655</td>\n",
       "      <td>3014836</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60656</td>\n",
       "      <td>3014841</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60657</td>\n",
       "      <td>3014844</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60658</td>\n",
       "      <td>3014849</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60659</td>\n",
       "      <td>3014852</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60660 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0 seqname   source feature  start    end  score strand  frame  \\\n",
       "0               0    chr1   HAVANA    gene  11869  14409    NaN      +      0   \n",
       "1              12    chr1   HAVANA    gene  14404  29570    NaN      -      0   \n",
       "2              25    chr1  ENSEMBL    gene  17369  17436    NaN      -      0   \n",
       "3              28    chr1   HAVANA    gene  29554  31109    NaN      +      0   \n",
       "4              36    chr1  ENSEMBL    gene  30366  30503    NaN      +      0   \n",
       "...           ...     ...      ...     ...    ...    ...    ...    ...    ...   \n",
       "60655     3014836    chrM  ENSEMBL    gene  14149  14673    NaN      -      0   \n",
       "60656     3014841    chrM  ENSEMBL    gene  14674  14742    NaN      -      0   \n",
       "60657     3014844    chrM  ENSEMBL    gene  14747  15887    NaN      +      0   \n",
       "60658     3014849    chrM  ENSEMBL    gene  15888  15953    NaN      +      0   \n",
       "60659     3014852    chrM  ENSEMBL    gene  15956  16023    NaN      -      0   \n",
       "\n",
       "                 gene_id  ... transcript_type transcript_name  \\\n",
       "0      ENSG00000223972.5  ...             NaN             NaN   \n",
       "1      ENSG00000227232.5  ...             NaN             NaN   \n",
       "2      ENSG00000278267.1  ...             NaN             NaN   \n",
       "3      ENSG00000243485.5  ...             NaN             NaN   \n",
       "4      ENSG00000284332.1  ...             NaN             NaN   \n",
       "...                  ...  ...             ...             ...   \n",
       "60655  ENSG00000198695.2  ...             NaN             NaN   \n",
       "60656  ENSG00000210194.1  ...             NaN             NaN   \n",
       "60657  ENSG00000198727.2  ...             NaN             NaN   \n",
       "60658  ENSG00000210195.2  ...             NaN             NaN   \n",
       "60659  ENSG00000210196.2  ...             NaN             NaN   \n",
       "\n",
       "       transcript_support_level         tag havana_transcript  exon_number  \\\n",
       "0                           NaN         NaN               NaN          NaN   \n",
       "1                           NaN         NaN               NaN          NaN   \n",
       "2                           NaN         NaN               NaN          NaN   \n",
       "3                           NaN  ncRNA_host               NaN          NaN   \n",
       "4                           NaN         NaN               NaN          NaN   \n",
       "...                         ...         ...               ...          ...   \n",
       "60655                       NaN         NaN               NaN          NaN   \n",
       "60656                       NaN         NaN               NaN          NaN   \n",
       "60657                       NaN         NaN               NaN          NaN   \n",
       "60658                       NaN         NaN               NaN          NaN   \n",
       "60659                       NaN         NaN               NaN          NaN   \n",
       "\n",
       "       exon_id  ont  protein_id ccdsid  \n",
       "0          NaN  NaN         NaN    NaN  \n",
       "1          NaN  NaN         NaN    NaN  \n",
       "2          NaN  NaN         NaN    NaN  \n",
       "3          NaN  NaN         NaN    NaN  \n",
       "4          NaN  NaN         NaN    NaN  \n",
       "...        ...  ...         ...    ...  \n",
       "60655      NaN  NaN         NaN    NaN  \n",
       "60656      NaN  NaN         NaN    NaN  \n",
       "60657      NaN  NaN         NaN    NaN  \n",
       "60658      NaN  NaN         NaN    NaN  \n",
       "60659      NaN  NaN         NaN    NaN  \n",
       "\n",
       "[60660 rows x 26 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "feature_table['seqname'] = feature_table.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binary_find (array, target):\n",
    "    left = 0\n",
    "    right = len(array)-1\n",
    "    while (right-left)!=1:\n",
    "        if array[(right+left)//2] > target:\n",
    "            right = (right+left)//2\n",
    "        elif array[(right+left)//2] < target: \n",
    "            left = (right+left)//2\n",
    "    return right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "chr_names=['chrM', 'chrX', 'chrY']\n",
    "for i in range(1, 23):\n",
    "    chr_names.append('chr%s' % (i))\n",
    "chr_names.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': [0, 60659],\n",
       " 'chr10': [60659, 60659],\n",
       " 'chr11': [60659, 60659],\n",
       " 'chr12': [60659, 60659],\n",
       " 'chr13': [60659, 60659],\n",
       " 'chr14': [60659, 60659],\n",
       " 'chr15': [60659, 60659],\n",
       " 'chr16': [60659, 60659],\n",
       " 'chr17': [60659, 60659],\n",
       " 'chr18': [60659, 60659],\n",
       " 'chr19': [60659, 60659],\n",
       " 'chr2': [60659, 60659],\n",
       " 'chr20': [60659, 60659],\n",
       " 'chr21': [60659, 60659],\n",
       " 'chr22': [60659, 60659],\n",
       " 'chr3': [60659, 60659],\n",
       " 'chr4': [60659, 60659],\n",
       " 'chr5': [60659, 60659],\n",
       " 'chr6': [60659, 60659],\n",
       " 'chr7': [60659, 60659],\n",
       " 'chr8': [60659, 60659],\n",
       " 'chr9': [60659, 60659],\n",
       " 'chrM': [60659, 60659],\n",
       " 'chrX': [60659, 60659],\n",
       " 'chrY': [60659, 60659]}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates={}  \n",
    "for i in range(len(chr_names)):\n",
    "    a = binary_find(feature_table['seqname'], chr_names[i]+' ')\n",
    "    coordinates[chr_names[i]] = [binary_find(feature_table['seqname'], chr_names[i-1]+' '), binary_find(feature_table['seqname'], chr_names[i]+' ')]                \n",
    "coordinates[chr_names[0]][0] = 0\n",
    "coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': [0, 60659],\n",
       " 'chr10': [60659, 60659],\n",
       " 'chr11': [60659, 60659],\n",
       " 'chr12': [60659, 60659],\n",
       " 'chr13': [60659, 60659],\n",
       " 'chr14': [60659, 60659],\n",
       " 'chr15': [60659, 60659],\n",
       " 'chr16': [60659, 60659],\n",
       " 'chr17': [60659, 60659],\n",
       " 'chr18': [60659, 60659],\n",
       " 'chr19': [60659, 60659],\n",
       " 'chr2': [60659, 60659],\n",
       " 'chr20': [60659, 60659],\n",
       " 'chr21': [60659, 60659],\n",
       " 'chr22': [60659, 60659],\n",
       " 'chr3': [60659, 60659],\n",
       " 'chr4': [60659, 60659],\n",
       " 'chr5': [60659, 60659],\n",
       " 'chr6': [60659, 60659],\n",
       " 'chr7': [60659, 60659],\n",
       " 'chr8': [60659, 60659],\n",
       " 'chr9': [60659, 60659],\n",
       " 'chrM': [60659, 60659],\n",
       " 'chrX': [60659, 60659],\n",
       " 'chrY': [60659, 60659]}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordinates={}  \n",
    "for i in range(len(chr_names)):\n",
    "    a = binary_find(feature_table['seqname'], chr_names[i]+' ')\n",
    "    coordinates[chr_names[i]] = [binary_find(feature_table['seqname'], chr_names[i-1]+' '), binary_find(feature_table['seqname'], chr_names[i]+' ')]                \n",
    "coordinates[chr_names[0]][0] = 0\n",
    "coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "for chromosome in chr_names[:-1]:\n",
    "    feature_table[coordinates[chromosome][0]:coordinates[chromosome][1]].to_csv('~/Documents/CourseWork/genomes/Annotation/chr_features/%s_genes.tsv' % (chromosome) , sep='\\t', index=False)\n",
    "feature_table[coordinates[chr_names[-1]][0]:coordinates[chr_names[-1]][1]+1].to_csv('~/Documents/CourseWork/genomes/Annotation/chr_features/%s_genes.tsv' % (chr_names[-1]), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
