{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/dask/config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  data = yaml.load(f.read()) or {}\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/dask/dataframe/utils.py:13: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/distributed/config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  defaults = yaml.load(f)\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:529: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:530: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
      "/home/pitikovegor/anaconda/anaconda/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:535: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import ray\n",
    "from multiprocessing import Process, Pipe\n",
    "import xgboost as xgb\n",
    "import numpy as np \n",
    "import sklearn\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.datasets import make_classification\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import (RandomTreesEmbedding, RandomForestClassifier, GradientBoostingClassifier)\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.pipeline import make_pipeline\n",
    "import tensorflow\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import precision_score\n",
    "import matplotlib.gridspec as gridspec\n",
    "from kpal.klib import Profile\n",
    "from multiprocessing import Queue\n",
    "from multiprocessing import Process\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from multiprocessing import Pool, freeze_support\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Info about paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'K562'\n",
    "mod_id = 'GSM4041595_K562_rep2.hg38'\n",
    "k = 4\n",
    "window = 600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cont_data_path = f'/home/pitikovegor/CourseWork/genomes/reds_experiments/{cell_type}/all_parsed/'\n",
    "sequence_path = '/home/pitikovegor/CourseWork/genomes/chr_sequence'\n",
    "window = 500\n",
    "\n",
    "chr_names=['chrX', 'chrY']\n",
    "for i in range(1, 23):\n",
    "    chr_names.append('chr%d' % (i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Count kmers and make sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {'pval_trash_good': 0.05,\n",
    "          'pval_trash_bad': 0.1,\n",
    "          'rna_count_trash':  900,\n",
    "          'cont_count_trash': 0,\n",
    "          'zscore_trash_good': 0,\n",
    "          'zscore_trash_bad_max': 1,\n",
    "          'zscore_trash_bad_min': -1,\n",
    "          'clus_count_rna' : 10,\n",
    "          'clus_count_dna' : 10,\n",
    "          'class_trash_good': [''],\n",
    "          'class_trash_bad': ['protein_coding']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract k-mers\n",
    "def seq_cutter(sequence, position, chain, window = 500):\n",
    "    if chain=='+':\n",
    "        return sequence[max(0, position[0]+1-window) : min(position[1]+1+window, len(sequence))]\n",
    "    else:\n",
    "        return sequence[max(0, position[0]+1-window) : min(position[1]+1+window, len(sequence))].reverse_complement()\n",
    "\n",
    "def open_fasta (name):\n",
    "    for record in SeqIO.parse(name, \"fasta\"): # It's a FUCKING GENERATOR!!!!!!! I need in 1st - all of the rest is alternatively\n",
    "        chr_seq = record.seq\n",
    "        del record\n",
    "        return chr_seq\n",
    "    \n",
    "#def k_mer_counter (k, sequence):\n",
    " #   k_mers_count = max(len(sequence) - k + 1, 1)\n",
    "  #  sequence = [sequence]\n",
    "   # k_mers = Profile.from_sequences(sequence, k).counts\n",
    "  #  k_mers = np.multiply(k_mers, np.array([1/k_mers_count]*len(k_mers)))\n",
    " #   return k_mers\n",
    "\n",
    "\n",
    "def k_mer_counter (k, sequence):\n",
    "    k_mers_count = max(len(sequence) - k + 1, 1)\n",
    "    sequence = [sequence]\n",
    "    all_kmers_freqs = []\n",
    "    for i in range(k):\n",
    "        k_mer = Profile.from_sequences(list(sequence), i).counts\n",
    "        k_mer_count = max(len(sequence) - i + 1, 1)\n",
    "        k_mer = np.multiply(k_mer, np.array([1/k_mer_count]*len(k_mer)))\n",
    "        all_kmers_freqs.append(k_mer)\n",
    "    k_mers = Profile.from_sequences(sequence, k).counts\n",
    "    k_mers = np.multiply(k_mers, np.array([1/k_mers_count]*len(k_mers)))\n",
    "    for i in range(len(all_kmers_freqs)):\n",
    "        k_count = 4**(i+1)\n",
    "        for j in range(len(k_mers)//k_count):\n",
    "            k_mers[j*k_count:(j+1)*k_count]\n",
    "    return k_mers\n",
    "\n",
    "\n",
    "def sequence_process_rna (fasta_seq, k, input_table, window = 500):\n",
    "    chr_seq = open_fasta(fasta_seq)\n",
    "    input_table = input_table.to_numpy()\n",
    "    all_kmers = []\n",
    "    for cont in input_table:\n",
    "        seq = seq_cutter(chr_seq, (cont[1], cont[2]), cont[3], window = window)\n",
    "        seq = str(seq)\n",
    "        k_mer_num = list(k_mer_counter(k, seq))\n",
    "        all_kmers.append(k_mer_num)\n",
    "    if all_kmers:\n",
    "        all_kmers = np.array(all_kmers)\n",
    "        input_table = np.concatenate((input_table, all_kmers), axis=1)\n",
    "        input_table = input_table[~np.all(input_table[:, 4:] == 0, axis=1)]\n",
    "        return pd.DataFrame(input_table)\n",
    "    else:\n",
    "        return pd.DataFrame(columns = list(range(5+4**k)))\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "def make_sets (contact_data_table, params):\n",
    "    contacts = pd.read_table(contact_data_table, sep='\\t')\n",
    "    contacts = contacts.astype({\"pval\": float, 'contacts_count': int, 'rnaseq_count': int, 'zscore': float})\n",
    "    contacts['class'] = 0\n",
    "    conditions = [\n",
    "        #(contacts['pval'] <= params['pval_trash_good']) &\n",
    "        (contacts['rnaseq_count'] >= params['rna_count_trash']) &\n",
    "        (contacts['zscore'] >= params['zscore_trash_good']) &\n",
    "        (contacts['contacts_count'] >= params['cont_count_trash']) &\n",
    "        (contacts['rna_chr']!= contacts['dna_chr']) &\n",
    "        (contacts['clus_count_rna'] >= params['clus_count_rna']) &\n",
    "        (contacts['clus_count_dna'] >= params['clus_count_dna'])\n",
    "     ]\n",
    "    values = [1]\n",
    "    contacts['class'] = pd.Series(np.select(conditions, values))\n",
    "    return contacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "def all_tables(chr_seq, contact_data_table, params, k, window = 500):\n",
    "    chr_cont = make_sets(contact_data_table, params)\n",
    "    print('%s is %s length' % (contact_data_table.split('/')[-1], len(chr_cont['class'])))\n",
    "    true_positive = chr_cont[(chr_cont['class']==1)]\n",
    "    true_positive = true_positive.sample(n = min(len(true_positive['class']), 100000))\n",
    "    true_negative = chr_cont[#(chr_cont['pval']>=params['pval_trash_bad']) & \n",
    "                             (chr_cont['dna_chr']!=chr_cont['rna_chr'])]\n",
    "    print('%s negative after pval %s length' % (contact_data_table.split('/')[-1], len(true_negative['class'])))\n",
    "    true_negative = true_negative[(true_negative['gene_type'] == 'protein_coding') & \n",
    "                                  (true_negative['clus_count_dna'] < 3)]\n",
    "    true_negative = true_negative.sample(n = min(len(true_negative['class']), len(true_positive['class'])*5))\n",
    "    del chr_cont\n",
    "    print('Start k-mers %s' % contact_data_table.split(\"/\")[-1])\n",
    "    kmers_positive = sequence_process_rna(chr_seq, k, true_positive[['id', 'rna_bgn', 'rna_end', 'rna_strand', 'class']], window = window)\n",
    "    kmers_negative = sequence_process_rna(chr_seq, k, true_negative[['id', 'rna_bgn', 'rna_end', 'rna_strand', 'class']], window = window)\n",
    "    print('From %s %s positive %s negative' % (contact_data_table.split(\"/\")[-1], len(true_positive['class']), len(true_negative['class'])))\n",
    "    return true_positive, true_negative, kmers_positive, kmers_negative  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "run_control": {
     "marked": false
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrY_all_table.tsv is 127 length\n",
      "chrY_all_table.tsv negative after pval 127 length\n",
      "Start k-mers chrY_all_table.tsv\n",
      "From chrY_all_table.tsv 0 positive 0 negative\n",
      "chr20_all_table.tsv is 586255 length\n",
      "chr20_all_table.tsv negative after pval 539390 length\n",
      "Start k-mers chr20_all_table.tsv\n",
      "chr13_all_table.tsv is 859896 length\n",
      "chr13_all_table.tsv negative after pval 744561 length\n",
      "Start k-mers chr13_all_table.tsv\n",
      "chr18_all_table.tsv is 747458 length\n",
      "chr18_all_table.tsv negative after pval 653191 length\n",
      "Start k-mers chr18_all_table.tsv\n",
      "chr21_all_table.tsv is 767169 length\n",
      "chr17_all_table.tsv is 1212269 length\n",
      "chr21_all_table.tsv negative after pval 718237 length\n",
      "chr17_all_table.tsv negative after pval 1121865 length\n",
      "chr14_all_table.tsv is 1373433 length\n",
      "chr22_all_table.tsv is 829711 length\n",
      "Start k-mers chr21_all_table.tsv\n",
      "chr15_all_table.tsv is 857678 length\n",
      "chr22_all_table.tsv negative after pval 783959 length\n",
      "Start k-mers chr17_all_table.tsv\n",
      "chr15_all_table.tsv negative after pval 768162 length\n",
      "chrX_all_table.tsv is 749033 length\n",
      "chr14_all_table.tsv negative after pval 1191611 length\n",
      "chrX_all_table.tsv negative after pval 647771 length\n",
      "chr12_all_table.tsv is 1345035 length\n",
      "Start k-mers chr15_all_table.tsv\n",
      "Start k-mers chr22_all_table.tsv\n",
      "Start k-mers chrX_all_table.tsv\n",
      "Start k-mers chr14_all_table.tsv\n",
      "chr12_all_table.tsv negative after pval 1168928 length\n",
      "Start k-mers chr12_all_table.tsv\n",
      "chr16_all_table.tsv is 1043875 length\n",
      "chr16_all_table.tsv negative after pval 961248 length\n",
      "chr8_all_table.tsv is 1564331 length\n",
      "Start k-mers chr16_all_table.tsv\n",
      "chr19_all_table.tsv is 1002320 length\n",
      "chr10_all_table.tsv is 1212995 length\n",
      "chr8_all_table.tsv negative after pval 1303236 length\n",
      "chr19_all_table.tsv negative after pval 936760 length\n",
      "chr4_all_table.tsv is 1221053 length\n",
      "chr10_all_table.tsv negative after pval 1033674 length\n",
      "Start k-mers chr19_all_table.tsv\n",
      "Start k-mers chr8_all_table.tsv\n",
      "chr4_all_table.tsv negative after pval 995203 length\n",
      "Start k-mers chr10_all_table.tsv\n",
      "Start k-mers chr4_all_table.tsv\n",
      "chr6_all_table.tsv is 1916182 length\n",
      "chr6_all_table.tsv negative after pval 1571718 length\n",
      "Start k-mers chr6_all_table.tsv\n",
      "chr9_all_table.tsv is 1377911 length\n",
      "chr9_all_table.tsv negative after pval 1229573 length\n",
      "From chr20_all_table.tsv 189 positive 945 negative\n",
      "From chr19_all_table.tsv 35 positive 175 negative\n",
      "Start k-mers chr9_all_table.tsv\n",
      "chr3_all_table.tsv is 1948262 length\n",
      "chr3_all_table.tsv negative after pval 1597742 length\n",
      "chr5_all_table.tsv is 1762370 length\n",
      "Start k-mers chr3_all_table.tsv\n",
      "chr5_all_table.tsv negative after pval 1471821 length\n",
      "Start k-mers chr5_all_table.tsv\n",
      "chr7_all_table.tsv is 2209918 length\n",
      "chr7_all_table.tsv negative after pval 1843246 length\n",
      "From chrX_all_table.tsv 148 positive 740 negative\n",
      "Start k-mers chr7_all_table.tsv\n",
      "chr11_all_table.tsv is 2941452 length\n",
      "chr11_all_table.tsv negative after pval 2633283 length\n",
      "Start k-mers chr11_all_table.tsv\n",
      "chr2_all_table.tsv is 2943784 length\n",
      "chr2_all_table.tsv negative after pval 2282632 length\n",
      "Start k-mers chr2_all_table.tsv\n",
      "chr1_all_table.tsv is 3553932 length\n",
      "chr1_all_table.tsv negative after pval 2884596 length\n",
      "Start k-mers chr1_all_table.tsv\n",
      "From chr15_all_table.tsv 631 positive 3155 negative\n",
      "From chr12_all_table.tsv 1049 positive 5245 negative\n",
      "From chr10_all_table.tsv 1760 positive 8800 negative\n",
      "From chr17_all_table.tsv 1844 positive 9220 negative\n",
      "From chr18_all_table.tsv 2566 positive 12830 negative\n",
      "From chr13_all_table.tsv 3739 positive 18695 negative\n",
      "From chr6_all_table.tsv 4795 positive 23975 negative\n",
      "From chr22_all_table.tsv 6831 positive 34155 negative\n",
      "From chr21_all_table.tsv 5459 positive 27295 negative\n",
      "From chr3_all_table.tsv 5975 positive 29875 negative\n",
      "From chr16_all_table.tsv 6538 positive 32690 negative\n",
      "From chr8_all_table.tsv 9319 positive 46595 negative\n",
      "From chr9_all_table.tsv 9939 positive 49695 negative\n",
      "From chr5_all_table.tsv 14057 positive 70285 negative\n",
      "From chr1_all_table.tsv 12742 positive 63710 negative\n",
      "From chr2_all_table.tsv 18056 positive 90280 negative\n",
      "From chr4_all_table.tsv 25692 positive 128460 negative\n",
      "From chr14_all_table.tsv 31905 positive 159525 negative\n",
      "From chr7_all_table.tsv 98622 positive 493110 negative\n",
      "From chr11_all_table.tsv 100000 positive 500000 negative\n"
     ]
    }
   ],
   "source": [
    "all_proc = Pool(processes = 24)\n",
    "#all_data.close()\n",
    "all_data = all_proc.starmap(all_tables, [('%s/%s.fna' % (sequence_path, i), f'{cont_data_path}/{i}_all_table.tsv', params, k, window) for i in chr_names])\n",
    "all_proc.close()\n",
    "all_proc.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_contacts = pd.concat([i[0] for i in all_data] + [i[1] for i in all_data], ignore_index = True)\n",
    "all_kmers = pd.concat([i[2] for i in all_data] + [i[3] for i in all_data], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "      <th>256</th>\n",
       "      <th>257</th>\n",
       "      <th>258</th>\n",
       "      <th>259</th>\n",
       "      <th>260</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D00795:30:CA2UTANXX:1:1310:18334:18583</td>\n",
       "      <td>9539350</td>\n",
       "      <td>9541431</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0136452</td>\n",
       "      <td>0.00454841</td>\n",
       "      <td>0.00682261</td>\n",
       "      <td>0.00844704</td>\n",
       "      <td>0.00682261</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000324886</td>\n",
       "      <td>0.00779727</td>\n",
       "      <td>0.00454841</td>\n",
       "      <td>0.00389864</td>\n",
       "      <td>0.00844704</td>\n",
       "      <td>0.00974659</td>\n",
       "      <td>0.00454841</td>\n",
       "      <td>0.00649773</td>\n",
       "      <td>0.00844704</td>\n",
       "      <td>0.00747238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1117:8781:19203</td>\n",
       "      <td>74217251</td>\n",
       "      <td>74219273</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0152368</td>\n",
       "      <td>0.00596224</td>\n",
       "      <td>0.005631</td>\n",
       "      <td>0.00828089</td>\n",
       "      <td>0.005631</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000662471</td>\n",
       "      <td>0.00529977</td>\n",
       "      <td>0.00728718</td>\n",
       "      <td>0.00662471</td>\n",
       "      <td>0.00397483</td>\n",
       "      <td>0.00794965</td>\n",
       "      <td>0.00662471</td>\n",
       "      <td>0.00629347</td>\n",
       "      <td>0.00828089</td>\n",
       "      <td>0.00794965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D00795:30:CA2UTANXX:1:1207:2249:88194</td>\n",
       "      <td>26722780</td>\n",
       "      <td>26724810</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0066072</td>\n",
       "      <td>0.00231252</td>\n",
       "      <td>0.0066072</td>\n",
       "      <td>0.0066072</td>\n",
       "      <td>0.00528576</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00033036</td>\n",
       "      <td>0.00693756</td>\n",
       "      <td>0.0049554</td>\n",
       "      <td>0.00429468</td>\n",
       "      <td>0.00594648</td>\n",
       "      <td>0.00561612</td>\n",
       "      <td>0.011893</td>\n",
       "      <td>0.00759828</td>\n",
       "      <td>0.00925008</td>\n",
       "      <td>0.0181698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D00795:28:C99BVACXX:7:1108:19414:11801</td>\n",
       "      <td>26723792</td>\n",
       "      <td>26725835</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00789474</td>\n",
       "      <td>0.00361842</td>\n",
       "      <td>0.00625</td>\n",
       "      <td>0.00789474</td>\n",
       "      <td>0.00690789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000328947</td>\n",
       "      <td>0.00921053</td>\n",
       "      <td>0.00756579</td>\n",
       "      <td>0.00460526</td>\n",
       "      <td>0.00526316</td>\n",
       "      <td>0.00756579</td>\n",
       "      <td>0.0148026</td>\n",
       "      <td>0.0101974</td>\n",
       "      <td>0.0105263</td>\n",
       "      <td>0.0266447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D00795:30:CA2UTANXX:5:2203:19145:4481</td>\n",
       "      <td>74244493</td>\n",
       "      <td>74246513</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0129267</td>\n",
       "      <td>0.00364601</td>\n",
       "      <td>0.00331455</td>\n",
       "      <td>0.00994365</td>\n",
       "      <td>0.00729201</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000994365</td>\n",
       "      <td>0.0116009</td>\n",
       "      <td>0.00530328</td>\n",
       "      <td>0.00629765</td>\n",
       "      <td>0.0029831</td>\n",
       "      <td>0.0106066</td>\n",
       "      <td>0.0172357</td>\n",
       "      <td>0.0135897</td>\n",
       "      <td>0.0125953</td>\n",
       "      <td>0.0417633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171341</th>\n",
       "      <td>D00795:30:CA2UTANXX:1:1307:11995:99947</td>\n",
       "      <td>24621019</td>\n",
       "      <td>24623083</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0192747</td>\n",
       "      <td>0.00326691</td>\n",
       "      <td>0.00424698</td>\n",
       "      <td>0.00392029</td>\n",
       "      <td>0.00522705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000326691</td>\n",
       "      <td>0.00261352</td>\n",
       "      <td>0.00261352</td>\n",
       "      <td>0.00261352</td>\n",
       "      <td>0.00392029</td>\n",
       "      <td>0.00228683</td>\n",
       "      <td>0.000653381</td>\n",
       "      <td>0.00130676</td>\n",
       "      <td>0.00294022</td>\n",
       "      <td>0.000980072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171342</th>\n",
       "      <td>D00795:30:CA2UTANXX:1:2202:3479:33822</td>\n",
       "      <td>29288803</td>\n",
       "      <td>29290860</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0127701</td>\n",
       "      <td>0.00491159</td>\n",
       "      <td>0.00982318</td>\n",
       "      <td>0.00884086</td>\n",
       "      <td>0.00556647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000982318</td>\n",
       "      <td>0.00589391</td>\n",
       "      <td>0.00687623</td>\n",
       "      <td>0.00392927</td>\n",
       "      <td>0.00589391</td>\n",
       "      <td>0.00982318</td>\n",
       "      <td>0.0117878</td>\n",
       "      <td>0.00884086</td>\n",
       "      <td>0.0108055</td>\n",
       "      <td>0.0173543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171343</th>\n",
       "      <td>D00795:30:CA2UTANXX:5:1312:12454:38826</td>\n",
       "      <td>20539774</td>\n",
       "      <td>20541851</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0240729</td>\n",
       "      <td>0.00553025</td>\n",
       "      <td>0.00553025</td>\n",
       "      <td>0.0156148</td>\n",
       "      <td>0.00683149</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000650618</td>\n",
       "      <td>0.00422902</td>\n",
       "      <td>0.00585556</td>\n",
       "      <td>0.0035784</td>\n",
       "      <td>0.00585556</td>\n",
       "      <td>0.00585556</td>\n",
       "      <td>0.00748211</td>\n",
       "      <td>0.00585556</td>\n",
       "      <td>0.00845804</td>\n",
       "      <td>0.0208198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171344</th>\n",
       "      <td>D00795:30:CA2UTANXX:1:2303:11148:74521</td>\n",
       "      <td>18125413</td>\n",
       "      <td>18127509</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0126091</td>\n",
       "      <td>0.00226317</td>\n",
       "      <td>0.00484966</td>\n",
       "      <td>0.00484966</td>\n",
       "      <td>0.00258649</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00581959</td>\n",
       "      <td>0.00452635</td>\n",
       "      <td>0.00387973</td>\n",
       "      <td>0.00549628</td>\n",
       "      <td>0.00581959</td>\n",
       "      <td>0.00484966</td>\n",
       "      <td>0.00452635</td>\n",
       "      <td>0.00840608</td>\n",
       "      <td>0.0210152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171345</th>\n",
       "      <td>D00795:28:C99BVACXX:7:1114:12236:91423</td>\n",
       "      <td>19498774</td>\n",
       "      <td>19500812</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0062603</td>\n",
       "      <td>0.00461285</td>\n",
       "      <td>0.00197694</td>\n",
       "      <td>0.00230643</td>\n",
       "      <td>0.00560132</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000988468</td>\n",
       "      <td>0.00263591</td>\n",
       "      <td>0.00329489</td>\n",
       "      <td>0.00395387</td>\n",
       "      <td>0.00593081</td>\n",
       "      <td>0.00263591</td>\n",
       "      <td>0.000988468</td>\n",
       "      <td>0.00230643</td>\n",
       "      <td>0.00461285</td>\n",
       "      <td>0.00230643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2171346 rows × 261 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 0         1         2   3    \\\n",
       "0             D00795:30:CA2UTANXX:1:1310:18334:18583   9539350   9541431   +   \n",
       "1        M01460:89:000000000-AJJ2J:1:1117:8781:19203  74217251  74219273   -   \n",
       "2              D00795:30:CA2UTANXX:1:1207:2249:88194  26722780  26724810   +   \n",
       "3             D00795:28:C99BVACXX:7:1108:19414:11801  26723792  26725835   +   \n",
       "4              D00795:30:CA2UTANXX:5:2203:19145:4481  74244493  74246513   -   \n",
       "...                                              ...       ...       ...  ..   \n",
       "2171341       D00795:30:CA2UTANXX:1:1307:11995:99947  24621019  24623083   +   \n",
       "2171342        D00795:30:CA2UTANXX:1:2202:3479:33822  29288803  29290860   +   \n",
       "2171343       D00795:30:CA2UTANXX:5:1312:12454:38826  20539774  20541851   +   \n",
       "2171344       D00795:30:CA2UTANXX:1:2303:11148:74521  18125413  18127509   +   \n",
       "2171345       D00795:28:C99BVACXX:7:1114:12236:91423  19498774  19500812   +   \n",
       "\n",
       "        4           5           6           7           8           9    ...  \\\n",
       "0         1   0.0136452  0.00454841  0.00682261  0.00844704  0.00682261  ...   \n",
       "1         1   0.0152368  0.00596224    0.005631  0.00828089    0.005631  ...   \n",
       "2         1   0.0066072  0.00231252   0.0066072   0.0066072  0.00528576  ...   \n",
       "3         1  0.00789474  0.00361842     0.00625  0.00789474  0.00690789  ...   \n",
       "4         1   0.0129267  0.00364601  0.00331455  0.00994365  0.00729201  ...   \n",
       "...      ..         ...         ...         ...         ...         ...  ...   \n",
       "2171341   0   0.0192747  0.00326691  0.00424698  0.00392029  0.00522705  ...   \n",
       "2171342   0   0.0127701  0.00491159  0.00982318  0.00884086  0.00556647  ...   \n",
       "2171343   0   0.0240729  0.00553025  0.00553025   0.0156148  0.00683149  ...   \n",
       "2171344   0   0.0126091  0.00226317  0.00484966  0.00484966  0.00258649  ...   \n",
       "2171345   0   0.0062603  0.00461285  0.00197694  0.00230643  0.00560132  ...   \n",
       "\n",
       "                 251         252         253         254         255  \\\n",
       "0        0.000324886  0.00779727  0.00454841  0.00389864  0.00844704   \n",
       "1        0.000662471  0.00529977  0.00728718  0.00662471  0.00397483   \n",
       "2         0.00033036  0.00693756   0.0049554  0.00429468  0.00594648   \n",
       "3        0.000328947  0.00921053  0.00756579  0.00460526  0.00526316   \n",
       "4        0.000994365   0.0116009  0.00530328  0.00629765   0.0029831   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "2171341  0.000326691  0.00261352  0.00261352  0.00261352  0.00392029   \n",
       "2171342  0.000982318  0.00589391  0.00687623  0.00392927  0.00589391   \n",
       "2171343  0.000650618  0.00422902  0.00585556   0.0035784  0.00585556   \n",
       "2171344            0  0.00581959  0.00452635  0.00387973  0.00549628   \n",
       "2171345  0.000988468  0.00263591  0.00329489  0.00395387  0.00593081   \n",
       "\n",
       "                256          257         258         259          260  \n",
       "0        0.00974659   0.00454841  0.00649773  0.00844704   0.00747238  \n",
       "1        0.00794965   0.00662471  0.00629347  0.00828089   0.00794965  \n",
       "2        0.00561612     0.011893  0.00759828  0.00925008    0.0181698  \n",
       "3        0.00756579    0.0148026   0.0101974   0.0105263    0.0266447  \n",
       "4         0.0106066    0.0172357   0.0135897   0.0125953    0.0417633  \n",
       "...             ...          ...         ...         ...          ...  \n",
       "2171341  0.00228683  0.000653381  0.00130676  0.00294022  0.000980072  \n",
       "2171342  0.00982318    0.0117878  0.00884086   0.0108055    0.0173543  \n",
       "2171343  0.00585556   0.00748211  0.00585556  0.00845804    0.0208198  \n",
       "2171344  0.00581959   0.00484966  0.00452635  0.00840608    0.0210152  \n",
       "2171345  0.00263591  0.000988468  0.00230643  0.00461285   0.00230643  \n",
       "\n",
       "[2171346 rows x 261 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contacts_count</th>\n",
       "      <th>rnaseq_count</th>\n",
       "      <th>cont</th>\n",
       "      <th>rnaseq</th>\n",
       "      <th>chP</th>\n",
       "      <th>zscore</th>\n",
       "      <th>pval</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensg</th>\n",
       "      <th>rna_chr</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_bgn</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>ac</th>\n",
       "      <th>clus_count_rna</th>\n",
       "      <th>dna_chr</th>\n",
       "      <th>dna_bgn</th>\n",
       "      <th>dna_end</th>\n",
       "      <th>clus_count_dna</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19840</td>\n",
       "      <td>2879</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>2.054347</td>\n",
       "      <td>1.605305</td>\n",
       "      <td>0.108427</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>ENSG00000101849.17</td>\n",
       "      <td>chrX</td>\n",
       "      <td>...</td>\n",
       "      <td>9463295</td>\n",
       "      <td>9741037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000101849.17</td>\n",
       "      <td>18245</td>\n",
       "      <td>chr13</td>\n",
       "      <td>91126141</td>\n",
       "      <td>91128164</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24154</td>\n",
       "      <td>3527</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>2.048147</td>\n",
       "      <td>1.599703</td>\n",
       "      <td>0.109665</td>\n",
       "      <td>FTX</td>\n",
       "      <td>ENSG00000230590.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>...</td>\n",
       "      <td>73940435</td>\n",
       "      <td>74293574</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000230590.11</td>\n",
       "      <td>24504</td>\n",
       "      <td>chr13</td>\n",
       "      <td>91198941</td>\n",
       "      <td>91200965</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16510</td>\n",
       "      <td>4787</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>1.362301</td>\n",
       "      <td>0.979939</td>\n",
       "      <td>0.327116</td>\n",
       "      <td>VENTXP1</td>\n",
       "      <td>ENSG00000259849.2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>...</td>\n",
       "      <td>26557675</td>\n",
       "      <td>26881029</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>ENSG00000259849.2</td>\n",
       "      <td>15956</td>\n",
       "      <td>chr13</td>\n",
       "      <td>91010016</td>\n",
       "      <td>91012039</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16510</td>\n",
       "      <td>4787</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>1.362301</td>\n",
       "      <td>0.979939</td>\n",
       "      <td>0.327116</td>\n",
       "      <td>VENTXP1</td>\n",
       "      <td>ENSG00000259849.2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>...</td>\n",
       "      <td>26557675</td>\n",
       "      <td>26881029</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>ENSG00000259849.2</td>\n",
       "      <td>15956</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23231022</td>\n",
       "      <td>23233046</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>24154</td>\n",
       "      <td>3527</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>2.048147</td>\n",
       "      <td>1.599703</td>\n",
       "      <td>0.109665</td>\n",
       "      <td>FTX</td>\n",
       "      <td>ENSG00000230590.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>...</td>\n",
       "      <td>73940435</td>\n",
       "      <td>74293574</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000230590.11</td>\n",
       "      <td>24504</td>\n",
       "      <td>chr13</td>\n",
       "      <td>91273146</td>\n",
       "      <td>91275170</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171341</th>\n",
       "      <td>104</td>\n",
       "      <td>86</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.312439</td>\n",
       "      <td>0.031233</td>\n",
       "      <td>0.975084</td>\n",
       "      <td>AP000356.5</td>\n",
       "      <td>ENSG00000286070.2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>...</td>\n",
       "      <td>24556007</td>\n",
       "      <td>24628987</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000286070.2</td>\n",
       "      <td>1494</td>\n",
       "      <td>chr8</td>\n",
       "      <td>119659816</td>\n",
       "      <td>119661839</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171342</th>\n",
       "      <td>9128</td>\n",
       "      <td>9830</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.050303</td>\n",
       "      <td>-0.205647</td>\n",
       "      <td>0.837067</td>\n",
       "      <td>EWSR1</td>\n",
       "      <td>ENSG00000182944.18</td>\n",
       "      <td>chr22</td>\n",
       "      <td>...</td>\n",
       "      <td>29268009</td>\n",
       "      <td>29300525</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000182944.18</td>\n",
       "      <td>8820</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236947857</td>\n",
       "      <td>236949880</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171343</th>\n",
       "      <td>16800</td>\n",
       "      <td>5075</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>1.321302</td>\n",
       "      <td>0.942890</td>\n",
       "      <td>0.345737</td>\n",
       "      <td>MED15</td>\n",
       "      <td>ENSG00000099917.17</td>\n",
       "      <td>chr22</td>\n",
       "      <td>...</td>\n",
       "      <td>20495913</td>\n",
       "      <td>20587632</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000099917.17</td>\n",
       "      <td>15380</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140017779</td>\n",
       "      <td>140019802</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171344</th>\n",
       "      <td>2638</td>\n",
       "      <td>2068</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.367722</td>\n",
       "      <td>0.081188</td>\n",
       "      <td>0.935292</td>\n",
       "      <td>PEX26</td>\n",
       "      <td>ENSG00000215193.13</td>\n",
       "      <td>chr22</td>\n",
       "      <td>...</td>\n",
       "      <td>18077920</td>\n",
       "      <td>18131138</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000215193.13</td>\n",
       "      <td>2460</td>\n",
       "      <td>chr5</td>\n",
       "      <td>61976111</td>\n",
       "      <td>61978134</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171345</th>\n",
       "      <td>9658</td>\n",
       "      <td>6565</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.510372</td>\n",
       "      <td>0.210095</td>\n",
       "      <td>0.833594</td>\n",
       "      <td>CDC45</td>\n",
       "      <td>ENSG00000093009.11</td>\n",
       "      <td>chr22</td>\n",
       "      <td>...</td>\n",
       "      <td>19479457</td>\n",
       "      <td>19520612</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000093009.11</td>\n",
       "      <td>8802</td>\n",
       "      <td>chr7</td>\n",
       "      <td>19090677</td>\n",
       "      <td>19092701</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2171346 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         contacts_count  rnaseq_count      cont    rnaseq       chP    zscore  \\\n",
       "0                 19840          2879  0.000525  0.000067  2.054347  1.605305   \n",
       "1                 24154          3527  0.000640  0.000082  2.048147  1.599703   \n",
       "2                 16510          4787  0.000437  0.000112  1.362301  0.979939   \n",
       "3                 16510          4787  0.000437  0.000112  1.362301  0.979939   \n",
       "4                 24154          3527  0.000640  0.000082  2.048147  1.599703   \n",
       "...                 ...           ...       ...       ...       ...       ...   \n",
       "2171341             104            86  0.000003  0.000002  0.312439  0.031233   \n",
       "2171342            9128          9830  0.000242  0.000230  0.050303 -0.205647   \n",
       "2171343           16800          5075  0.000445  0.000119  1.321302  0.942890   \n",
       "2171344            2638          2068  0.000070  0.000048  0.367722  0.081188   \n",
       "2171345            9658          6565  0.000256  0.000154  0.510372  0.210095   \n",
       "\n",
       "             pval   gene_name                ensg rna_chr  ...  gene_bgn  \\\n",
       "0        0.108427       TBL1X  ENSG00000101849.17    chrX  ...   9463295   \n",
       "1        0.109665         FTX  ENSG00000230590.11    chrX  ...  73940435   \n",
       "2        0.327116     VENTXP1   ENSG00000259849.2    chrX  ...  26557675   \n",
       "3        0.327116     VENTXP1   ENSG00000259849.2    chrX  ...  26557675   \n",
       "4        0.109665         FTX  ENSG00000230590.11    chrX  ...  73940435   \n",
       "...           ...         ...                 ...     ...  ...       ...   \n",
       "2171341  0.975084  AP000356.5   ENSG00000286070.2   chr22  ...  24556007   \n",
       "2171342  0.837067       EWSR1  ENSG00000182944.18   chr22  ...  29268009   \n",
       "2171343  0.345737       MED15  ENSG00000099917.17   chr22  ...  20495913   \n",
       "2171344  0.935292       PEX26  ENSG00000215193.13   chr22  ...  18077920   \n",
       "2171345  0.833594       CDC45  ENSG00000093009.11   chr22  ...  19479457   \n",
       "\n",
       "         gene_end                         gene_type                  ac  \\\n",
       "0         9741037                    protein_coding  ENSG00000101849.17   \n",
       "1        74293574                            lncRNA  ENSG00000230590.11   \n",
       "2        26881029  transcribed_processed_pseudogene   ENSG00000259849.2   \n",
       "3        26881029  transcribed_processed_pseudogene   ENSG00000259849.2   \n",
       "4        74293574                            lncRNA  ENSG00000230590.11   \n",
       "...           ...                               ...                 ...   \n",
       "2171341  24628987                    protein_coding   ENSG00000286070.2   \n",
       "2171342  29300525                    protein_coding  ENSG00000182944.18   \n",
       "2171343  20587632                    protein_coding  ENSG00000099917.17   \n",
       "2171344  18131138                    protein_coding  ENSG00000215193.13   \n",
       "2171345  19520612                    protein_coding  ENSG00000093009.11   \n",
       "\n",
       "         clus_count_rna  dna_chr    dna_bgn    dna_end  clus_count_dna class  \n",
       "0                 18245    chr13   91126141   91128164              11     1  \n",
       "1                 24504    chr13   91198941   91200965              16     1  \n",
       "2                 15956    chr13   91010016   91012039              13     1  \n",
       "3                 15956    chr22   23231022   23233046              11     1  \n",
       "4                 24504    chr13   91273146   91275170              16     1  \n",
       "...                 ...      ...        ...        ...             ...   ...  \n",
       "2171341            1494     chr8  119659816  119661839               1     0  \n",
       "2171342            8820     chr1  236947857  236949880               1     0  \n",
       "2171343           15380     chr5  140017779  140019802               1     0  \n",
       "2171344            2460     chr5   61976111   61978134               1     0  \n",
       "2171345            8802     chr7   19090677   19092701               1     0  \n",
       "\n",
       "[2171346 rows x 24 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_contacts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_contacts = all_kmers[all_kmers[4]==1]\n",
    "good_contacts = good_contacts.sample(n = min(len(good_contacts[0]), 500000))\n",
    "bad_contacts = all_kmers[all_kmers[4]==0]\n",
    "bad_contacts = bad_contacts.sample(n = min(len(good_contacts[0])*10, len(bad_contacts[0])))\n",
    "all_set_contacts = good_contacts.append(bad_contacts, ignore_index = True)\n",
    "del good_contacts\n",
    "del bad_contacts\n",
    "all_set_contacts = all_set_contacts.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_train, data_test, label_train, label_test = train_test_split(all_set_contacts[:, 5:], all_set_contacts[:, 4], test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "dtrain = xgb.DMatrix(data_train, label=label_train)\n",
    "dtest = xgb.DMatrix(data_test, label=label_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "param = {\n",
    "    'max_depth': 1000,\n",
    "    'eta': 0.3, \n",
    "    'silent': 4, \n",
    "    'objective': 'multi:softprob',\n",
    "    'num_class': 2,\n",
    "    'nthread' : 10,\n",
    "    'num_parallel_tree': 4,\n",
    "    'booster': 'gbtree'\n",
    "    }\n",
    "num_boost_round = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "fit() got an unexpected keyword argument 'num_boost_round'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-93ce8e9d7f88>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mbst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mxgb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mXGBClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparam\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mbst\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtrain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_boost_round\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnum_boost_round\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mtrees\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0ms\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mbst\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrees\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mnum_boost_round\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: fit() got an unexpected keyword argument 'num_boost_round'"
     ]
    }
   ],
   "source": [
    "bst = xgb.XGBClassifier(param)\n",
    "bst.fit(dtrain, num_boost_round=num_boost_round)\n",
    "\n",
    "trees = [s for s in bst]\n",
    "assert len(trees) == num_boost_round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "?xgb.XGBClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "now = datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "q = 30\n",
    "name = 'xgb_clusters_%s_%s_%s' % (current_time, k, mod_id)\n",
    "bst.save_model('/Users/pitikov_egor/Documents/CourseWork/genomes/models/%s' % name)\n",
    "bst.dump_model('/Users/pitikov_egor/Documents/CourseWork/genomes/models/%s_dump' % name, '/Users/pitikov_egor/Documents/CourseWork/genomes/models/features_%s' % name)\n",
    "q+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = bst.predict(dtest)\n",
    "best_preds = np.asarray([np.argmax(line) for line in preds])\n",
    "precision_score(list(label_test), best_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fpr, tpr, thresholds = roc_curve(list(label_test), list(preds[:, 1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(1, figsize=(30, 30))\n",
    "plt.plot([0, 1], [0, 1], 'k--')\n",
    "plt.plot(fpr, tpr, label='epoch=%s, xgb'%(num_boost_round))\n",
    "plt.xlabel('False positive rate', fontsize=15)\n",
    "plt.ylabel('True positive rate', fontsize=15)\n",
    "plt.title('ROC curve xgb 4-mers', fontsize=30)\n",
    "plt.legend(loc='best', fontsize=20)\n",
    "plt.savefig('/Users/pitikov_egor/Documents/CourseWork/genomes/pictures/Models/%s_ROC.png' % name)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del all_set_contacts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Open and check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "modelname = 'xgb_trans_15:17:46_GSM4041595_K562_rep2.hg38'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "bst = xgb.Booster({'nthread': 10})\n",
    "bst.load_model('/Users/pitikov_egor/Documents/CourseWork/genomes/models/%s.model' % modelname) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "cell_type = 'K562'\n",
    "experiment = 'GSM4041593_K562_rep1.hg38'\n",
    "cont_data_path = '/Users/pitikov_egor/Documents/CourseWork/genomes/reds_experiments/%s/parsed_contacts_%s' % (cell_type, experiment)\n",
    "sequence_path = '/Users/pitikov_egor/Documents/CourseWork/genomes/chr_sequence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_test_sets (contact_data_table, redc_table):\n",
    "    contacts = pd.read_table(contact_data_table, sep='\\t')\n",
    "    contacts = contacts.astype({\"pval\": float, 'contacts_count': int, 'rnaseq_count': int, 'zscore': float})\n",
    "    redc = pd.read_table(redc_table, sep='\\t')\n",
    "    contacts = pd.merge(left=contacts, right=redc[['id', 'rna_chr', 'rna_bgn', 'rna_end', 'dna_bgn', 'dna_end', 'dna_chr', 'rna_strand', 'dna_strand']], left_on='id', right_on='id', how ='inner')\n",
    "    del redc\n",
    "    #contacts = contacts[contacts['rna_chr']!=contacts['dna_chr']]\n",
    "    contacts['class'] = 0\n",
    "    return contacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_test_tables(chr_seq, contact_data_table, redc_table, k):\n",
    "    chr_cont = make_test_sets(contact_data_table,  redc_table)\n",
    "    kmer = sequence_process_rna(chr_seq, k, chr_cont[['id', 'rna_bgn', 'rna_end', 'rna_strand', 'gene_type']])\n",
    "    return chr_cont, kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "all_proc_test = Pool(processes = 24)\n",
    "all_data = all_proc_test.starmap(all_test_tables, [('%s/%s.fna' % (sequence_path, i), '%s/%s_full_%s.pvalue.tab.gene.txt' % (cont_data_path, i, cell_type), '%s/%s_%s.tsv' % (cont_data_path, i, experiment), k) for i in chr_names])\n",
    "all_proc_test.close()\n",
    "all_proc_test.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_contacts = pd.concat([i[0] for i in all_data], ignore_index = True)\n",
    "all_kmers = pd.concat([i[1] for i in all_data], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "all_kmers = all_kmers.to_numpy()\n",
    "full_kmers = xgb.DMatrix(all_kmers[:, 5:])\n",
    "preds = bst.predict(full_kmers)\n",
    "best_preds = np.asarray([np.argmax(line) for line in preds])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_preds\n",
    "preds_to_type = np.concatenate((best_preds[:, np.newaxis], all_kmers[:, 4][:, np.newaxis], all_kmers[:, 0][:, np.newaxis]), axis=1)\n",
    "gene_types = list(set(all_kmers[:, 4]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miRNA have 0 specific and 54 nonspecific contacts\n",
      "TEC have 0 specific and 297 nonspecific contacts\n",
      "snRNA have 812 specific and 4480 nonspecific contacts\n",
      "transcribed_unitary_pseudogene have 2 specific and 1022 nonspecific contacts\n",
      "polymorphic_pseudogene have 0 specific and 12 nonspecific contacts\n",
      "scRNA have 0 specific and 1 nonspecific contacts\n",
      "transcribed_processed_pseudogene have 0 specific and 8470 nonspecific contacts\n",
      "processed_pseudogene have 0 specific and 516 nonspecific contacts\n",
      "transcribed_unprocessed_pseudogene have 0 specific and 5134 nonspecific contacts\n",
      "ribozyme have 0 specific and 748 nonspecific contacts\n",
      "unprocessed_pseudogene have 0 specific and 455 nonspecific contacts\n",
      "unitary_pseudogene have 0 specific and 5 nonspecific contacts\n",
      "snoRNA have 0 specific and 4915 nonspecific contacts\n",
      "IG_V_gene have 0 specific and 3 nonspecific contacts\n",
      "lncRNA have 2373 specific and 205347 nonspecific contacts\n",
      "protein_coding have 2790 specific and 1016469 nonspecific contacts\n",
      "scaRNA have 0 specific and 584 nonspecific contacts\n",
      "misc_RNA have 0 specific and 439 nonspecific contacts\n",
      "rRNA_pseudogene have 0 specific and 5 nonspecific contacts\n"
     ]
    }
   ],
   "source": [
    "means = {}\n",
    "for i in gene_types:\n",
    "    spec = np.count_nonzero((preds_to_type[:, 0] == 1) & (preds_to_type[:, 1] == i))\n",
    "    nonspec = np.count_nonzero((preds_to_type[:, 0] == 0) & (preds_to_type[:, 1] == i))\n",
    "    print('%s have %s specific and %s nonspecific contacts' % (i, spec, nonspec))\n",
    "    means[i] = [nonspec, spec]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA1YAAAZwCAYAAAB+kdTnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzcMW5c2YGG0XsHEsCknagrI2wqMLyB2sME2gfXoCVoDdqHgFlEeQHGJDLArMxkIgEd3Alsw4ZhN8X+JNYt+5yMhariHxAgPrz73lxrDQAAAH65/7r0AAAAgGsnrAAAACJhBQAAEAkrAACASFgBAABEzwqrOef/fK8hAAAAu/tXTfTqOV/yq1/96r+Px6PnswMAAP+p/u+fvfissPrtb387TqfTt5kDAABwZeac//vPXnePFQAAQCSsAAAAImEFAAAQPeseKwAAgO/lp59+Gg8PD+PLly+XnjJubm7G7e3teP369Ve9X1gBAABbeHh4GD/88MO4u7sbc86L7VhrjcfHx/Hw8DDevn37VZ9xFBAAANjCly9fxps3by4aVWOMMeccb968edaVM2EFAABs49JR9VfP3eEoIAAAsKW795++6fd9/vDum37f33PFCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJPBQQA4Fm+9ZPanut7PtkNfilhBQAAbOmaItpRQAAAgEhYAQAA21hrXXrCGOP5O4QVAACwhZubm/H4+HjxuFprjcfHx3Fzc/PVn3GPFQAAsIXb29vx8PAwzufzpaeMm5ubcXt7+9XvF1YAAMAWXr9+Pd6+fXvpGb+Io4AAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAADRk2E157yfc57mnKfz+fwSmwAAAK7Kk2G11vq41jqutY6Hw+ElNgEAAFwVRwEBAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAESvLj0AAIC/uXv/6dITxucP7y49Aa6OK1YAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAANGTYTXnvJ9znuacp/P5/BKbAAAArsqTYbXW+rjWOq61jofD4SU2AQAAXBVHAQEAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAANGTYTXnvJ9znuacp/P5/BKbAAAArsqTYbXW+rjWOq61jofD4SU2AQAAXBVHAQEAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACB6MqzmnPdzztOc83Q+n19iEwAAwFV5MqzWWh/XWse11vFwOLzEJgAAgKviKCAAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQvbr0APZw9/7TRX//5w/vLvr7AQCgcMUKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAA0ZNhNee8n3Oe5pyn8/n8EpsAAACuypNhtdb6uNY6rrWOh8PhJTYBAABcFUcBAQAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIHp16QHwNe7ef7r0hPH5w7tLTwAAYFNzrfXzb5jzfoxx/5cffzfG+MP3HnUBP44x/nTpET9j931j7L9x931j7L9x931j7L9x931j7L9x931j7L9x931j7L9x931j7L9x931j7L/Rvsv5zVrr8I8vPhlW/wnmnKe11vHSO/6V3feNsf/G3feNsf/G3feNsf/G3feNsf/G3feNsf/G3feNsf/G3feNsf/G3feNsf9G+/bjHisAAIBIWAEAAETC6s8+XnrAE3bfN8b+G3ffN8b+G3ffN8b+G3ffN8b+G3ffN8b+G3ffN8b+G3ffN8b+G3ffN8b+G+3bjHusAAAAIlesAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAED06jlv/vHHH9fd3d13mgIAALC33//+939aax3+8fVnhdXd3d04nU7fbhUAAMAVmXP+8Z+97iggAABAJKwAAAAiYQUAABA96x4rAACA7+Wnn34aDw8P48uXL5eeMm5ubsbt7e14/fr1V71fWAEAAFt4eHgYP/zww7i7uxtzzovtWGuNx8fH8fDwMN6+fftVn3EUEAAA2MKXL1/GmzdvLhpVY4wx5xxv3rx51pUzYQUAAGzj0lH1V8/d4SggAACwpbv3n77p933+8O6bft/fc8UKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAIk8FBADgWb71k9qe63s+2Q1+KWEFAABs6Zoi2lFAAACASFgBAADbWGtdesIY4/k7hBUAALCFm5ub8fj4ePG4WmuNx8fHcXNz89WfefIeqznn/Rjjfowxfv3rX//ydQAAAD/j9vZ2PDw8jPP5fOkp4+bmZtze3n71+58Mq7XWxzHGxzHGOB6Pe1yXAwAA/u28fv16vH379tIzfhFHAQEAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARE+G1Zzzfs55mnOezufzS2wCAAC4Kk+G1Vrr41rruNY6Hg6Hl9gEAABwVRwFBAAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAoifDas55P+c8zTlP5/P5JTYBAABclSfDaq31ca11XGsdD4fDS2wCAAC4Ko4CAgAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAA0ZNhNee8n3Oe5pyn8/n8EpsAAACuypNhtdb6uNY6rrWOh8PhJTYBAABcFUcBAQAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAAKJXlx4AAMDf3L3/dOkJ4/OHd5eeAFfHFSsAAIDoybCac97POU9zztP5fH6JTQAAAFflybBaa31cax3XWsfD4fASmwAAAK6Ko4AAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIHp16QEAf3X3/tNFf//nD+8u+vsBgOslrAD+jYhTLu3Sf4Nj+DsELkNYMca4/D9C/wQBALhmc63182+Y836Mcf+XH383xvjD9x51AT+OMf506RE/Y/d9Y+y/cfd9Y+y/cfd9Y+y/cfd9Y+y/cfd9Y+y/cfd9Y+y/cfd9Y+y/cfd9Y+y/0b7L+c1a6/CPLz4ZVv8J5pyntdbx0jv+ld33jbH/xt33jbH/xt33jbH/xt33jbH/xt33jbH/xt33jbH/xt33jbH/xt33jbH/Rvv246mAAAAAkbACAACIhNWffbz0gCfsvm+M/Tfuvm+M/Tfuvm+M/Tfuvm+M/Tfuvm+M/Tfuvm+M/Tfuvm+M/Tfuvm+M/Tfatxn3WAEAAESuWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAA/r+9O8Zt9EjQMFwFuAEljtrMBKw62gvwFr6HzuAj+Ay6h29BH2AzL6CMo7wBBzXBzgKDmZ1ha982WXQ/T9YEJX7R33j1F0kgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAHrwX4wAAAtySURBVETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACD67j1P/uGHH9bT09MfNAUAAGBvv/7661/WWod/fPxdYfX09DROp9PXWwUAAHBH5pz//X897iggAABAJKwAAAAiYQUAABC96z1WAAAAf5Tff/99vL6+js+fP996ynh4eBiPj4/jw4cPX/R8YQUAAGzh9fV1fP/99+Pp6WnMOW+2Y6013t7exuvr6/j06dMX/YyjgAAAwBY+f/48Pn78eNOoGmOMOef4+PHju+6cCSsAAGAbt46q//XeHY4CAgAAW3r66Zev+vt++/nHr/r7/p47VgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgMjHrQPb+Nofqfpef+RHsAIA73dP/ze7YwUAABAJKwAAYBtrrVtPGGO8f4ewAgAAtvDw8DDe3t5uHldrrfH29jYeHh6++Ge8xwoAANjC4+PjeH19Hefz+dZTxsPDw3h8fPzi5wsrAABgCx8+fBifPn269Yz/F0cBAQAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACA6GJYzTmf55ynOefpfD5fYxMAAMBduRhWa62XtdZxrXU8HA7X2AQAAHBXHAUEAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAACii2E153yec57mnKfz+XyNTQAAAHflYlittV7WWse11vFwOFxjEwAAwF1xFBAAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAADRxbCacz7POU9zztP5fL7GJgAAgLtyMazWWi9rreNa63g4HK6xCQAA4K44CggAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIi+u/UAAL6ep59+uenr//bzjzd9fQC4FXesAAAAImEFAAAQCSsAAIBIWAEAAEQXw2rO+TznPM05T+fz+RqbAAAA7srFsFprvay1jmut4+FwuMYmAACAu+IoIAAAQOR7rACAr+bW36U2hu9TA27DHSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACA6GJYzTmf55ynOefpfD5fYxMAAMBduRhWa62XtdZxrXU8HA7X2AQAAHBXHAUEAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAdDGs5pzPc87TnPN0Pp+vsQkAAOCuXAyrtdbLWuu41joeDodrbAIAALgrjgICAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAADRxbCacz7POU9zztP5fL7GJgAAgLtyMazWWi9rreNa63g4HK6xCQAA4K44CggAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAAkbACAACIhBUAAEAkrAAAACJhBQAAEAkrAACASFgBAABEwgoAACASVgAAAJGwAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAgElYAAACRsAIAAIiEFQAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAEQXw2rO+TznPM05T+fz+RqbAAAA7srFsFprvay1jmut4+FwuMYmAACAu+IoIAAAQCSsAAAAImEFAAAQCSsAAIBIWAEAAETCCgAAIBJWAAAA0Xe3HgDAt+Ppp19u+vq//fzjTV8fgD8vd6wAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAg8qmAAAD8qdz6E0jH8Cmk3yJ3rAAAACJ3rAC+0K3/AuqvnwCwL3esAAAAImEFAAAQOQoIAH9z6+OeYzjyCd+KW19vXGu+PnesAAAAImEFAAAQOQoIAMC7OMYG/0xYcRdufQEfw0UcAIB/TVhdgSgAAIA/N++xAgAAiIQVAABAJKwAAAAiYQUAABAJKwAAgEhYAQAARMIKAAAg8j1WAAAb8f2XcJ/+FGF16wuQiw8AAHzbHAUEAACI5lrr3z9hzucxxvPf/vmfY4z/+qNH3cAPY4y/3HrEv7H7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH232jf7fzHWuvwjw9eDKtvwZzztNY63nrHv7L7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH237j7vjH232jffhwFBAAAiIQVAABAJKz+x8utB1yw+74x9t+4+74x9t+4+74x9t+4+74x9t+4+74x9t+4+74x9t+4+74x9t+4+74x9t9o32a8xwoAACByxwoAACASVgAAAJGwAgAAiIQVAABAJKwAAACivwJOhY0D5mrPqQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x2160 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.rcParams['axes.titlesize'] = 8\n",
    "fig = plt.figure(1, figsize=(15, 30))\n",
    "gs = gridspec.GridSpec(3, 2)\n",
    "ax0 = fig.add_subplot(gs[0, :])\n",
    "spec = ax0.bar(np.arange(len(means.keys())), [means[i][1] for i in means.keys()], label='specific')\n",
    "ax0.set_ylabel('Number')\n",
    "ax0.set_title('Specific by classes', fontsize=20)\n",
    "ax0.set_xticks(np.arange(len(means.keys()))) \n",
    "ax0.set_xticklabels(means.keys(), rotation=45,  ha=\"right\", position=(0, 0))\n",
    "ax0.legend()\n",
    "ax1 = fig.add_subplot(gs[1, :])\n",
    "nonspec = ax1.bar(np.arange(len(means.keys())), [means[i][0] for i in means.keys()], label='nonspecific')\n",
    "ax1.set_ylabel('Number')\n",
    "ax1.set_title('Nonspecific by classes', fontsize=20)\n",
    "ax1.set_xticks(np.arange(len(means.keys())))\n",
    "ax1.set_xticklabels(means.keys(), rotation=45,  ha=\"right\", position=(0, 0))\n",
    "ax1.legend()\n",
    "ax2 = fig.add_subplot(gs[2, :])\n",
    "nonspec = ax2.bar(np.arange(len(means.keys())), [means[i][0]/(means[i][1]+1) for i in means.keys()], label='nonspecific/specific')\n",
    "ax2.set_ylabel('Number')\n",
    "ax2.set_title('Nonspecific/specific by classes', fontsize=20)\n",
    "ax2.set_xticks(np.arange(len(means.keys())))\n",
    "ax2.set_xticklabels(means.keys(), rotation=45,  ha=\"right\", position=(0, 0))\n",
    "ax2.legend()\n",
    "plt.subplots_adjust(wspace=2, hspace=1)\n",
    "plt.savefig('/Users/pitikov_egor/Documents/CourseWork/genomes/pictures/Models/%s_coninfo.png' % modelname)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# return genes with specefic and nonespecific contacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds_to_type = pd.DataFrame(preds_to_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds_to_type = pd.merge(left=preds_to_type, right=all_contacts[['id', 'gene', 'gene_ac']], left_on=2, right_on='id', how ='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>id_x</th>\n",
       "      <th>gene_x</th>\n",
       "      <th>gene_ac_x</th>\n",
       "      <th>id_y</th>\n",
       "      <th>gene_y</th>\n",
       "      <th>gene_ac_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483489</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483490</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483491</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483492</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483493</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1483494 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0               1                                             2  \\\n",
       "0        0  protein_coding  M01460:89:000000000-AJJ2J:1:1101:18527:12404   \n",
       "1        0  protein_coding   M01460:89:000000000-AJJ2J:1:1101:27924:7073   \n",
       "2        0  protein_coding  M01460:89:000000000-AJJ2J:1:1102:11189:24626   \n",
       "3        0  protein_coding  M01460:89:000000000-AJJ2J:1:1102:11949:19953   \n",
       "4        0  protein_coding   M01460:89:000000000-AJJ2J:1:1102:17115:4112   \n",
       "...     ..             ...                                           ...   \n",
       "1483489  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:20963:14307   \n",
       "1483490  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:22798:23265   \n",
       "1483491  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:26090:21638   \n",
       "1483492  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:28153:15083   \n",
       "1483493  0  protein_coding   M01460:89:000000000-AJJ2J:1:2119:8126:20356   \n",
       "\n",
       "                                                 id_x gene_x  \\\n",
       "0        M01460:89:000000000-AJJ2J:1:1101:18527:12404  ABCB7   \n",
       "1         M01460:89:000000000-AJJ2J:1:1101:27924:7073  ABCB7   \n",
       "2        M01460:89:000000000-AJJ2J:1:1102:11189:24626  ABCB7   \n",
       "3        M01460:89:000000000-AJJ2J:1:1102:11949:19953  ABCB7   \n",
       "4         M01460:89:000000000-AJJ2J:1:1102:17115:4112  ABCB7   \n",
       "...                                               ...    ...   \n",
       "1483489  M01460:89:000000000-AJJ2J:1:2119:20963:14307  ZNRF3   \n",
       "1483490  M01460:89:000000000-AJJ2J:1:2119:22798:23265  ZNRF3   \n",
       "1483491  M01460:89:000000000-AJJ2J:1:2119:26090:21638  ZNRF3   \n",
       "1483492  M01460:89:000000000-AJJ2J:1:2119:28153:15083  ZNRF3   \n",
       "1483493   M01460:89:000000000-AJJ2J:1:2119:8126:20356  ZNRF3   \n",
       "\n",
       "                  gene_ac_x                                          id_y  \\\n",
       "0        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1101:18527:12404   \n",
       "1        ENSG00000131269.19   M01460:89:000000000-AJJ2J:1:1101:27924:7073   \n",
       "2        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1102:11189:24626   \n",
       "3        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1102:11949:19953   \n",
       "4        ENSG00000131269.19   M01460:89:000000000-AJJ2J:1:1102:17115:4112   \n",
       "...                     ...                                           ...   \n",
       "1483489  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:20963:14307   \n",
       "1483490  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:22798:23265   \n",
       "1483491  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:26090:21638   \n",
       "1483492  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:28153:15083   \n",
       "1483493  ENSG00000183579.16   M01460:89:000000000-AJJ2J:1:2119:8126:20356   \n",
       "\n",
       "        gene_y           gene_ac_y  \n",
       "0        ABCB7  ENSG00000131269.19  \n",
       "1        ABCB7  ENSG00000131269.19  \n",
       "2        ABCB7  ENSG00000131269.19  \n",
       "3        ABCB7  ENSG00000131269.19  \n",
       "4        ABCB7  ENSG00000131269.19  \n",
       "...        ...                 ...  \n",
       "1483489  ZNRF3  ENSG00000183579.16  \n",
       "1483490  ZNRF3  ENSG00000183579.16  \n",
       "1483491  ZNRF3  ENSG00000183579.16  \n",
       "1483492  ZNRF3  ENSG00000183579.16  \n",
       "1483493  ZNRF3  ENSG00000183579.16  \n",
       "\n",
       "[1483494 rows x 9 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds_to_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>id_x</th>\n",
       "      <th>gene_x</th>\n",
       "      <th>gene_ac_x</th>\n",
       "      <th>id_y</th>\n",
       "      <th>gene_y</th>\n",
       "      <th>gene_ac_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:18527:12404</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1101:27924:7073</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11189:24626</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:11949:19953</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:1102:17115:4112</td>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENSG00000131269.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483489</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:20963:14307</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483490</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:22798:23265</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483491</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:26090:21638</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483492</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:28153:15083</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1483493</td>\n",
       "      <td>0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "      <td>M01460:89:000000000-AJJ2J:1:2119:8126:20356</td>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>ENSG00000183579.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1483494 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0               1                                             2  \\\n",
       "0        0  protein_coding  M01460:89:000000000-AJJ2J:1:1101:18527:12404   \n",
       "1        0  protein_coding   M01460:89:000000000-AJJ2J:1:1101:27924:7073   \n",
       "2        0  protein_coding  M01460:89:000000000-AJJ2J:1:1102:11189:24626   \n",
       "3        0  protein_coding  M01460:89:000000000-AJJ2J:1:1102:11949:19953   \n",
       "4        0  protein_coding   M01460:89:000000000-AJJ2J:1:1102:17115:4112   \n",
       "...     ..             ...                                           ...   \n",
       "1483489  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:20963:14307   \n",
       "1483490  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:22798:23265   \n",
       "1483491  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:26090:21638   \n",
       "1483492  0  protein_coding  M01460:89:000000000-AJJ2J:1:2119:28153:15083   \n",
       "1483493  0  protein_coding   M01460:89:000000000-AJJ2J:1:2119:8126:20356   \n",
       "\n",
       "                                                 id_x gene_x  \\\n",
       "0        M01460:89:000000000-AJJ2J:1:1101:18527:12404  ABCB7   \n",
       "1         M01460:89:000000000-AJJ2J:1:1101:27924:7073  ABCB7   \n",
       "2        M01460:89:000000000-AJJ2J:1:1102:11189:24626  ABCB7   \n",
       "3        M01460:89:000000000-AJJ2J:1:1102:11949:19953  ABCB7   \n",
       "4         M01460:89:000000000-AJJ2J:1:1102:17115:4112  ABCB7   \n",
       "...                                               ...    ...   \n",
       "1483489  M01460:89:000000000-AJJ2J:1:2119:20963:14307  ZNRF3   \n",
       "1483490  M01460:89:000000000-AJJ2J:1:2119:22798:23265  ZNRF3   \n",
       "1483491  M01460:89:000000000-AJJ2J:1:2119:26090:21638  ZNRF3   \n",
       "1483492  M01460:89:000000000-AJJ2J:1:2119:28153:15083  ZNRF3   \n",
       "1483493   M01460:89:000000000-AJJ2J:1:2119:8126:20356  ZNRF3   \n",
       "\n",
       "                  gene_ac_x                                          id_y  \\\n",
       "0        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1101:18527:12404   \n",
       "1        ENSG00000131269.19   M01460:89:000000000-AJJ2J:1:1101:27924:7073   \n",
       "2        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1102:11189:24626   \n",
       "3        ENSG00000131269.19  M01460:89:000000000-AJJ2J:1:1102:11949:19953   \n",
       "4        ENSG00000131269.19   M01460:89:000000000-AJJ2J:1:1102:17115:4112   \n",
       "...                     ...                                           ...   \n",
       "1483489  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:20963:14307   \n",
       "1483490  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:22798:23265   \n",
       "1483491  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:26090:21638   \n",
       "1483492  ENSG00000183579.16  M01460:89:000000000-AJJ2J:1:2119:28153:15083   \n",
       "1483493  ENSG00000183579.16   M01460:89:000000000-AJJ2J:1:2119:8126:20356   \n",
       "\n",
       "        gene_y           gene_ac_y  \n",
       "0        ABCB7  ENSG00000131269.19  \n",
       "1        ABCB7  ENSG00000131269.19  \n",
       "2        ABCB7  ENSG00000131269.19  \n",
       "3        ABCB7  ENSG00000131269.19  \n",
       "4        ABCB7  ENSG00000131269.19  \n",
       "...        ...                 ...  \n",
       "1483489  ZNRF3  ENSG00000183579.16  \n",
       "1483490  ZNRF3  ENSG00000183579.16  \n",
       "1483491  ZNRF3  ENSG00000183579.16  \n",
       "1483492  ZNRF3  ENSG00000183579.16  \n",
       "1483493  ZNRF3  ENSG00000183579.16  \n",
       "\n",
       "[1483494 rows x 9 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds_to_type = preds_to_type.drop_duplicates()\n",
    "preds_to_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1194791"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(preds_to_type[2]))\n",
    "#There are contacts mapped on 2 and more genes - and I don't know, what's correct gene..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds_to_type['gene'] = preds_to_type['gene_x']\n",
    "preds_to_type['gene_ac'] = preds_to_type['gene_ac_x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene_preds = preds_to_type[['gene', 'gene_ac', 0, 1,]]\n",
    "gene_preds = gene_preds.rename(columns={0: 'class', 1: 'type'})\n",
    "gene_preds = gene_preds.groupby(gene_preds.columns.tolist()[::-1]).size()\n",
    "gene_preds = pd.DataFrame(gene_preds)\n",
    "gene_preds.reset_index(inplace=True)\n",
    "gene_preds = gene_preds.rename(columns={0: 'number'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       0\n",
       "3463    1\n",
       "Name: class, dtype: int64"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_preds['positive'] = 0\n",
    "gene_preds['negative'] = 0\n",
    "gene_preds.loc[gene_preds[gene_preds['class']==1].index, 'positive'] = gene_preds[gene_preds['class']==1].loc[:, 'number']\n",
    "gene_preds.loc[gene_preds[gene_preds['class']==0].index, 'negative'] = gene_preds[gene_preds['class']==0].loc[:, 'number']\n",
    "gene_preds.pop('class').drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>gene_ac</th>\n",
       "      <th>gene</th>\n",
       "      <th>number</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>ENSG00000211643.2</td>\n",
       "      <td>IGLV5-52</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>ENSG00000286129.1</td>\n",
       "      <td>AC245060.7</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSG00000048405.10</td>\n",
       "      <td>ZNF800</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSG00000051825.15</td>\n",
       "      <td>MPHOSPH9</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSG00000070366.14</td>\n",
       "      <td>SMG6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16401</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000283131.1</td>\n",
       "      <td>MTND2P41</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16402</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000283491.1</td>\n",
       "      <td>AC017104.4</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16403</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000284707.2</td>\n",
       "      <td>AC079781.5</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16404</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000285725.1</td>\n",
       "      <td>AC004967.2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16405</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000286073.1</td>\n",
       "      <td>AL138900.2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16406 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         type             gene_ac        gene  number  \\\n",
       "0                   IG_V_gene   ENSG00000211643.2    IGLV5-52       6   \n",
       "1                   IG_V_gene   ENSG00000286129.1  AC245060.7       6   \n",
       "2                         TEC  ENSG00000048405.10      ZNF800     112   \n",
       "3                         TEC  ENSG00000051825.15    MPHOSPH9       4   \n",
       "4                         TEC  ENSG00000070366.14        SMG6       2   \n",
       "...                       ...                 ...         ...     ...   \n",
       "16401  unprocessed_pseudogene   ENSG00000283131.1    MTND2P41       1   \n",
       "16402  unprocessed_pseudogene   ENSG00000283491.1  AC017104.4       7   \n",
       "16403  unprocessed_pseudogene   ENSG00000284707.2  AC079781.5       8   \n",
       "16404  unprocessed_pseudogene   ENSG00000285725.1  AC004967.2       6   \n",
       "16405  unprocessed_pseudogene   ENSG00000286073.1  AL138900.2       2   \n",
       "\n",
       "       positive  negative  \n",
       "0             0         6  \n",
       "1             0         6  \n",
       "2             0       112  \n",
       "3             0         4  \n",
       "4             0         2  \n",
       "...         ...       ...  \n",
       "16401         0         1  \n",
       "16402         0         7  \n",
       "16403         0         8  \n",
       "16404         0         6  \n",
       "16405         0         2  \n",
       "\n",
       "[16406 rows x 6 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_preds = gene_preds.sort_values(by = ['positive'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_preds = gene_preds.reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>gene_ac</th>\n",
       "      <th>gene</th>\n",
       "      <th>number</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>ENSG00000211643.2</td>\n",
       "      <td>IGLV5-52</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000163374.19</td>\n",
       "      <td>YY1AP1</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000163376.11</td>\n",
       "      <td>KBTBD8</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000163378.14</td>\n",
       "      <td>EOGT</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000163382.12</td>\n",
       "      <td>NAXE</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16401</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSG00000133316.15</td>\n",
       "      <td>WDR74</td>\n",
       "      <td>1624</td>\n",
       "      <td>1624</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16402</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>1624</td>\n",
       "      <td>1624</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16403</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000133316.15</td>\n",
       "      <td>WDR74</td>\n",
       "      <td>1624</td>\n",
       "      <td>1624</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16404</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000222328.1</td>\n",
       "      <td>RNU2-2P</td>\n",
       "      <td>1624</td>\n",
       "      <td>1624</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16405</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000188511.12</td>\n",
       "      <td>C22orf34</td>\n",
       "      <td>2025</td>\n",
       "      <td>2025</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16406 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 type             gene_ac      gene  number  positive  \\\n",
       "0           IG_V_gene   ENSG00000211643.2  IGLV5-52       6         0   \n",
       "1      protein_coding  ENSG00000163374.19    YY1AP1     111         0   \n",
       "2      protein_coding  ENSG00000163376.11    KBTBD8      24         0   \n",
       "3      protein_coding  ENSG00000163378.14      EOGT      24         0   \n",
       "4      protein_coding  ENSG00000163382.12      NAXE       4         0   \n",
       "...               ...                 ...       ...     ...       ...   \n",
       "16401           snRNA  ENSG00000133316.15     WDR74    1624      1624   \n",
       "16402           snRNA   ENSG00000222328.1   RNU2-2P    1624      1624   \n",
       "16403  protein_coding  ENSG00000133316.15     WDR74    1624      1624   \n",
       "16404  protein_coding   ENSG00000222328.1   RNU2-2P    1624      1624   \n",
       "16405          lncRNA  ENSG00000188511.12  C22orf34    2025      2025   \n",
       "\n",
       "       negative  \n",
       "0             6  \n",
       "1           111  \n",
       "2            24  \n",
       "3            24  \n",
       "4             4  \n",
       "...         ...  \n",
       "16401         0  \n",
       "16402         0  \n",
       "16403         0  \n",
       "16404         0  \n",
       "16405         0  \n",
       "\n",
       "[16406 rows x 6 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_preds.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalised = pd.read_table('/Users/pitikov_egor/Documents/CourseWork/genomes/normalisation_data/K562.pvalue.tab.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_info = pd.merge(left=normalised, right=gene_preds, left_on='gene_name', right_on='gene', how ='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_un</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>ensg</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contacts_count</th>\n",
       "      <th>rnaseq_count</th>\n",
       "      <th>...</th>\n",
       "      <th>rnaseq</th>\n",
       "      <th>chP</th>\n",
       "      <th>zscore</th>\n",
       "      <th>pval</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_ac</th>\n",
       "      <th>gene</th>\n",
       "      <th>number</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58347718</td>\n",
       "      <td>58355455</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000268895.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>122</td>\n",
       "      <td>122</td>\n",
       "      <td>...</td>\n",
       "      <td>2.876159e-06</td>\n",
       "      <td>0.124387</td>\n",
       "      <td>-0.138701</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000268895.6</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8822621</td>\n",
       "      <td>8887001</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>1.169170e-07</td>\n",
       "      <td>0.124387</td>\n",
       "      <td>-0.138701</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>33306766</td>\n",
       "      <td>33321098</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184389.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>2.104506e-07</td>\n",
       "      <td>1.538081</td>\n",
       "      <td>1.138781</td>\n",
       "      <td>0.254794</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000184389.9</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>chr12</td>\n",
       "      <td>53307456</td>\n",
       "      <td>53324864</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000094914.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>396</td>\n",
       "      <td>1380</td>\n",
       "      <td>...</td>\n",
       "      <td>3.229248e-05</td>\n",
       "      <td>-1.122240</td>\n",
       "      <td>-1.265213</td>\n",
       "      <td>0.205795</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000094914.14</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AACS</td>\n",
       "      <td>chr12</td>\n",
       "      <td>125065434</td>\n",
       "      <td>125143333</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000081760.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AACS</td>\n",
       "      <td>2520</td>\n",
       "      <td>1045</td>\n",
       "      <td>...</td>\n",
       "      <td>2.445904e-05</td>\n",
       "      <td>1.004069</td>\n",
       "      <td>0.656223</td>\n",
       "      <td>0.511681</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000081760.17</td>\n",
       "      <td>AACS</td>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16467</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52842511</td>\n",
       "      <td>52894998</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>7511</td>\n",
       "      <td>2400</td>\n",
       "      <td>...</td>\n",
       "      <td>5.614355e-05</td>\n",
       "      <td>1.265004</td>\n",
       "      <td>0.892016</td>\n",
       "      <td>0.372384</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16468</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52726453</td>\n",
       "      <td>52827336</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>2497</td>\n",
       "      <td>2130</td>\n",
       "      <td>...</td>\n",
       "      <td>4.983003e-05</td>\n",
       "      <td>0.283286</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>0.996100</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16469</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143381295</td>\n",
       "      <td>143391111</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>902</td>\n",
       "      <td>314</td>\n",
       "      <td>...</td>\n",
       "      <td>7.365772e-06</td>\n",
       "      <td>1.177537</td>\n",
       "      <td>0.812977</td>\n",
       "      <td>0.416231</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16470</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>4004445</td>\n",
       "      <td>4143030</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>4553</td>\n",
       "      <td>3343</td>\n",
       "      <td>...</td>\n",
       "      <td>7.819410e-05</td>\n",
       "      <td>0.433225</td>\n",
       "      <td>0.140381</td>\n",
       "      <td>0.888359</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>125</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16471</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>77562416</td>\n",
       "      <td>77683419</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>13957</td>\n",
       "      <td>3629</td>\n",
       "      <td>...</td>\n",
       "      <td>8.488176e-05</td>\n",
       "      <td>1.471207</td>\n",
       "      <td>1.078352</td>\n",
       "      <td>0.280877</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>458</td>\n",
       "      <td>0</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16472 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name_un gene_chr  gene_start   gene_end gene_strand  \\\n",
       "0         A1BG-AS1    chr19    58347718   58355455           +   \n",
       "1            A2ML1    chr12     8822621    8887001           +   \n",
       "2          A3GALT2     chr1    33306766   33321098           -   \n",
       "3             AAAS    chr12    53307456   53324864           -   \n",
       "4             AACS    chr12   125065434  125143333           +   \n",
       "...            ...      ...         ...        ...         ...   \n",
       "16467       ZYG11A     chr1    52842511   52894998           +   \n",
       "16468       ZYG11B     chr1    52726453   52827336           +   \n",
       "16469          ZYX     chr7   143381295  143391111           +   \n",
       "16470        ZZEF1    chr17     4004445    4143030           -   \n",
       "16471         ZZZ3     chr1    77562416   77683419           -   \n",
       "\n",
       "                     ensg       gene_type gene_name  contacts_count  \\\n",
       "0       ENSG00000268895.6          lncRNA  A1BG-AS1             122   \n",
       "1      ENSG00000166535.20  protein_coding     A2ML1               4   \n",
       "2       ENSG00000184389.9  protein_coding   A3GALT2              36   \n",
       "3      ENSG00000094914.14  protein_coding      AAAS             396   \n",
       "4      ENSG00000081760.17  protein_coding      AACS            2520   \n",
       "...                   ...             ...       ...             ...   \n",
       "16467  ENSG00000203995.10  protein_coding    ZYG11A            7511   \n",
       "16468  ENSG00000162378.13  protein_coding    ZYG11B            2497   \n",
       "16469  ENSG00000159840.16  protein_coding       ZYX             902   \n",
       "16470  ENSG00000074755.15  protein_coding     ZZEF1            4553   \n",
       "16471  ENSG00000036549.13  protein_coding      ZZZ3           13957   \n",
       "\n",
       "       rnaseq_count  ...        rnaseq       chP    zscore      pval  \\\n",
       "0               122  ...  2.876159e-06  0.124387 -0.138701  0.889687   \n",
       "1                 4  ...  1.169170e-07  0.124387 -0.138701  0.889687   \n",
       "2                 8  ...  2.104506e-07  1.538081  1.138781  0.254794   \n",
       "3              1380  ...  3.229248e-05 -1.122240 -1.265213  0.205795   \n",
       "4              1045  ...  2.445904e-05  1.004069  0.656223  0.511681   \n",
       "...             ...  ...           ...       ...       ...       ...   \n",
       "16467          2400  ...  5.614355e-05  1.265004  0.892016  0.372384   \n",
       "16468          2130  ...  4.983003e-05  0.283286  0.004888  0.996100   \n",
       "16469           314  ...  7.365772e-06  1.177537  0.812977  0.416231   \n",
       "16470          3343  ...  7.819410e-05  0.433225  0.140381  0.888359   \n",
       "16471          3629  ...  8.488176e-05  1.471207  1.078352  0.280877   \n",
       "\n",
       "                 type             gene_ac      gene number  positive  negative  \n",
       "0              lncRNA   ENSG00000268895.6  A1BG-AS1      9         0         9  \n",
       "1      protein_coding  ENSG00000166535.20     A2ML1      1         0         1  \n",
       "2      protein_coding   ENSG00000184389.9   A3GALT2      1         0         1  \n",
       "3      protein_coding  ENSG00000094914.14      AAAS     12         0        12  \n",
       "4      protein_coding  ENSG00000081760.17      AACS    108         0       108  \n",
       "...               ...                 ...       ...    ...       ...       ...  \n",
       "16467  protein_coding  ENSG00000203995.10    ZYG11A    261         0       261  \n",
       "16468  protein_coding  ENSG00000162378.13    ZYG11B     85         0        85  \n",
       "16469  protein_coding  ENSG00000159840.16       ZYX     25         0        25  \n",
       "16470  protein_coding  ENSG00000074755.15     ZZEF1    125         0       125  \n",
       "16471  protein_coding  ENSG00000036549.13      ZZZ3    458         0       458  \n",
       "\n",
       "[16472 rows x 21 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_info['con_per_mol'] = all_info.loc[:, 'contacts_count']/all_info.loc[:, 'rnaseq_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_un</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>ensg</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contacts_count</th>\n",
       "      <th>rnaseq_count</th>\n",
       "      <th>...</th>\n",
       "      <th>chP</th>\n",
       "      <th>zscore</th>\n",
       "      <th>pval</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_ac</th>\n",
       "      <th>gene</th>\n",
       "      <th>number</th>\n",
       "      <th>positive</th>\n",
       "      <th>negative</th>\n",
       "      <th>con_per_mol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58347718</td>\n",
       "      <td>58355455</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000268895.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>122</td>\n",
       "      <td>122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.124387</td>\n",
       "      <td>-0.138701</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000268895.6</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8822621</td>\n",
       "      <td>8887001</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.124387</td>\n",
       "      <td>-0.138701</td>\n",
       "      <td>0.889687</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>33306766</td>\n",
       "      <td>33321098</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000184389.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>1.538081</td>\n",
       "      <td>1.138781</td>\n",
       "      <td>0.254794</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000184389.9</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>chr12</td>\n",
       "      <td>53307456</td>\n",
       "      <td>53324864</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000094914.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>396</td>\n",
       "      <td>1380</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.122240</td>\n",
       "      <td>-1.265213</td>\n",
       "      <td>0.205795</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000094914.14</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0.286957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AACS</td>\n",
       "      <td>chr12</td>\n",
       "      <td>125065434</td>\n",
       "      <td>125143333</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000081760.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AACS</td>\n",
       "      <td>2520</td>\n",
       "      <td>1045</td>\n",
       "      <td>...</td>\n",
       "      <td>1.004069</td>\n",
       "      <td>0.656223</td>\n",
       "      <td>0.511681</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000081760.17</td>\n",
       "      <td>AACS</td>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>108</td>\n",
       "      <td>2.411483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16467</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52842511</td>\n",
       "      <td>52894998</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>7511</td>\n",
       "      <td>2400</td>\n",
       "      <td>...</td>\n",
       "      <td>1.265004</td>\n",
       "      <td>0.892016</td>\n",
       "      <td>0.372384</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>261</td>\n",
       "      <td>3.129583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16468</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52726453</td>\n",
       "      <td>52827336</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>2497</td>\n",
       "      <td>2130</td>\n",
       "      <td>...</td>\n",
       "      <td>0.283286</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>0.996100</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "      <td>1.172300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16469</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143381295</td>\n",
       "      <td>143391111</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>902</td>\n",
       "      <td>314</td>\n",
       "      <td>...</td>\n",
       "      <td>1.177537</td>\n",
       "      <td>0.812977</td>\n",
       "      <td>0.416231</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2.872611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16470</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>4004445</td>\n",
       "      <td>4143030</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>4553</td>\n",
       "      <td>3343</td>\n",
       "      <td>...</td>\n",
       "      <td>0.433225</td>\n",
       "      <td>0.140381</td>\n",
       "      <td>0.888359</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>125</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>1.361950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16471</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>77562416</td>\n",
       "      <td>77683419</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>13957</td>\n",
       "      <td>3629</td>\n",
       "      <td>...</td>\n",
       "      <td>1.471207</td>\n",
       "      <td>1.078352</td>\n",
       "      <td>0.280877</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>458</td>\n",
       "      <td>0</td>\n",
       "      <td>458</td>\n",
       "      <td>3.845963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16472 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name_un gene_chr  gene_start   gene_end gene_strand  \\\n",
       "0         A1BG-AS1    chr19    58347718   58355455           +   \n",
       "1            A2ML1    chr12     8822621    8887001           +   \n",
       "2          A3GALT2     chr1    33306766   33321098           -   \n",
       "3             AAAS    chr12    53307456   53324864           -   \n",
       "4             AACS    chr12   125065434  125143333           +   \n",
       "...            ...      ...         ...        ...         ...   \n",
       "16467       ZYG11A     chr1    52842511   52894998           +   \n",
       "16468       ZYG11B     chr1    52726453   52827336           +   \n",
       "16469          ZYX     chr7   143381295  143391111           +   \n",
       "16470        ZZEF1    chr17     4004445    4143030           -   \n",
       "16471         ZZZ3     chr1    77562416   77683419           -   \n",
       "\n",
       "                     ensg       gene_type gene_name  contacts_count  \\\n",
       "0       ENSG00000268895.6          lncRNA  A1BG-AS1             122   \n",
       "1      ENSG00000166535.20  protein_coding     A2ML1               4   \n",
       "2       ENSG00000184389.9  protein_coding   A3GALT2              36   \n",
       "3      ENSG00000094914.14  protein_coding      AAAS             396   \n",
       "4      ENSG00000081760.17  protein_coding      AACS            2520   \n",
       "...                   ...             ...       ...             ...   \n",
       "16467  ENSG00000203995.10  protein_coding    ZYG11A            7511   \n",
       "16468  ENSG00000162378.13  protein_coding    ZYG11B            2497   \n",
       "16469  ENSG00000159840.16  protein_coding       ZYX             902   \n",
       "16470  ENSG00000074755.15  protein_coding     ZZEF1            4553   \n",
       "16471  ENSG00000036549.13  protein_coding      ZZZ3           13957   \n",
       "\n",
       "       rnaseq_count  ...       chP    zscore      pval            type  \\\n",
       "0               122  ...  0.124387 -0.138701  0.889687          lncRNA   \n",
       "1                 4  ...  0.124387 -0.138701  0.889687  protein_coding   \n",
       "2                 8  ...  1.538081  1.138781  0.254794  protein_coding   \n",
       "3              1380  ... -1.122240 -1.265213  0.205795  protein_coding   \n",
       "4              1045  ...  1.004069  0.656223  0.511681  protein_coding   \n",
       "...             ...  ...       ...       ...       ...             ...   \n",
       "16467          2400  ...  1.265004  0.892016  0.372384  protein_coding   \n",
       "16468          2130  ...  0.283286  0.004888  0.996100  protein_coding   \n",
       "16469           314  ...  1.177537  0.812977  0.416231  protein_coding   \n",
       "16470          3343  ...  0.433225  0.140381  0.888359  protein_coding   \n",
       "16471          3629  ...  1.471207  1.078352  0.280877  protein_coding   \n",
       "\n",
       "                  gene_ac      gene number positive  negative  con_per_mol  \n",
       "0       ENSG00000268895.6  A1BG-AS1      9        0         9     1.000000  \n",
       "1      ENSG00000166535.20     A2ML1      1        0         1     1.000000  \n",
       "2       ENSG00000184389.9   A3GALT2      1        0         1     4.500000  \n",
       "3      ENSG00000094914.14      AAAS     12        0        12     0.286957  \n",
       "4      ENSG00000081760.17      AACS    108        0       108     2.411483  \n",
       "...                   ...       ...    ...      ...       ...          ...  \n",
       "16467  ENSG00000203995.10    ZYG11A    261        0       261     3.129583  \n",
       "16468  ENSG00000162378.13    ZYG11B     85        0        85     1.172300  \n",
       "16469  ENSG00000159840.16       ZYX     25        0        25     2.872611  \n",
       "16470  ENSG00000074755.15     ZZEF1    125        0       125     1.361950  \n",
       "16471  ENSG00000036549.13      ZZZ3    458        0       458     3.845963  \n",
       "\n",
       "[16472 rows x 22 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract k-mers\n",
    "def seq_cutter(sequence, position, chain, window = 500):\n",
    "    if chain=='+':\n",
    "        return sequence[max(0, position[0]+1-window) : min(position[1]+1+window, len(sequence))]\n",
    "    else:\n",
    "        return sequence[max(0, position[0]+1-window) : min(position[1]+1+window, len(sequence))].reverse_complement()\n",
    "\n",
    "def open_fasta (name):\n",
    "    for record in SeqIO.parse(name, \"fasta\"): # It's a FUCKING GENERATOR!!!!!!! I need in 1st - all of the rest is alternatively\n",
    "        chr_seq = record.seq\n",
    "        del record\n",
    "        return chr_seq\n",
    "\n",
    "def normalisation_kmers (k_list):\n",
    "    for i in range(math.log(len(k_list), 4)):\n",
    "        for j in range(len(k_list)//4**i):\n",
    "\n",
    "def k_mer_counter (k, sequence):\n",
    "    k_mers_count = max(len(sequence) - k + 1, 1)\n",
    "    sequence = [sequence]\n",
    "    all_kmers_freqs = []\n",
    "    for i in range(k):\n",
    "        k_mer = Profile.from_sequences(sequence, k).counts\n",
    "        k_mer = np.multiply(k_mers, np.array([1/k_mers_count]*len(k_mers)))\n",
    "        all_kmers_freqs.append(k_mer)\n",
    "    k_mers = Profile.from_sequences(sequence, k).counts\n",
    "    k_mers = np.multiply(k_mers, np.array([1/k_mers_count]*len(k_mers)))\n",
    "    for i in range(len(all_kmers_freqs)):\n",
    "        k_count = 4**(i+1)\n",
    "        for j in range(len(k_mers)//k_count):\n",
    "            k_mers[j*k_count:(j+1)*k_count]\n",
    "    return k_mers\n",
    "\n",
    "\n",
    "def k_mer_counter (k, sequence):\n",
    "    k_mers_count = max(len(sequence) - k + 1, 1)\n",
    "    sequence = [sequence]\n",
    "    k_mers = Profile.from_sequences(sequence, k).counts\n",
    "    k_mers = np.multiply(k_mers, np.array([1/k_mers_count]*len(k_mers)))\n",
    "    return k_mers\n",
    "\n",
    "\n",
    "def sequence_process_rna (fasta_seq, k, input_table, window = 500):\n",
    "    chr_seq = open_fasta(fasta_seq)\n",
    "    input_table = input_table.to_numpy()\n",
    "    all_kmers = []\n",
    "    for cont in input_table:\n",
    "        seq = seq_cutter(chr_seq, (cont[1], cont[2]), cont[3], window = window)\n",
    "        seq = str(seq)\n",
    "        k_mer_num = list(k_mer_counter(k, seq))\n",
    "        all_kmers.append(k_mer_num)\n",
    "    if all_kmers:\n",
    "        all_kmers = np.array(all_kmers)\n",
    "        input_table = np.concatenate((input_table, all_kmers), axis=1)\n",
    "        input_table = input_table[~np.all(input_table[:, 4:] == 0, axis=1)]\n",
    "        return pd.DataFrame(input_table)\n",
    "    else:\n",
    "        return pd.DataFrame(columns = list(range(5+4**k)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['a']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list('a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
